### DISCUSSION TITLE
Basic functions as synthetic data

### DISCUSSION THREAD

**Discussion by chadcwilliams** [AutoResearch/autora#546] [Ideas]
Building the tutorials, I use an experimentalist and a theorist from AutoRA, but I end up creating my own experiment runner of a sine wave because I have felt (for the sake of a tutorial) the synthetic data we have it more complicated than I want it to be (my ideal situation is that I wouldn't need to describe the data to our users). I am wondering if we see warrant in adding basic synthetic data just for benchmarking + tutorials, etc. 

I am talking super basic like sine waves with noise, exponential with noise, etc. 

For example:
```
def run_experiment(conditions: pd.DataFrame, added_noise: float = 0.5):
    x = conditions["x"]
    y = np.sin(x) + np.random.normal(0, added_noise, size=x.shape)
    observations = conditions.assign(y = y)
    return observations
```

EDIT: This is what I was thinking:

```
import sympy as sp
def equation_runner(expr: str, conditions: pd.DataFrame, added_noise = 0.1):
    equation = sp.simplify(expr)
    x = conditions["x"]
    
    y= [(equation.evalf(subs={equation.args[0]:xi}) + np.random.normal(0, added_noise, size=1))[0] for xi in x]
    return conditions.assign(y=y)

equation_runner('sin(x)', pd.DataFrame({'x':[0,1,3]}))
```

**Comment by chadcwilliams** [comment]
Here's an example as to why I think it might help. In the following, the experimental runner has to first be defined as a function but the others don't.

![Screenshot_2023-08-21 16_51_55](https://github.com/AutoResearch/autora/assets/26373111/25e93da1-af6e-4f59-9603-903a9a02bc8d)

**Reply by younesStrittmatter** [reply]
You could use the equation runner that uses any sympy expression as input here:  https://github.com/AutoResearch/autora-synthetic-abstract-equation.

**Reply by chadcwilliams** [reply]
Thanks! This is a step in the right direction, but still seems a bit clunky. Here's a comparison of the two:

```
#Function Method
def run_experiment(conditions: pd.DataFrame, added_noise: float = 0.5):
    x = conditions["x"]
    y = np.sin(x) + np.random.normal(0, added_noise, size=x.shape)
    observations = conditions.assign(y = y)
    return observations

#Equation Experiment Method
exp_iv = IV(name="x", value_range=(0, 2 * np.pi), allowed_values=np.linspace(0, 2 * np.pi, 30))
exp_dv = DV(name="y", type=ValueType.REAL)
sine_runner = equation_experiment(sp.simplify('sin(x)'),[exp_iv],exp_dv)
```

I'm basically hoping for a function that could work like this: `function('sin(x)')`. Perhaps it can work as a wrapper to your method? Something like this:

```
def equation_wrapper(expr: string, variables: VariableCollection):
    iv = [IV(variables.independent_variables[0])]
    dv = DV(variables.dependent_variables[0])
    return (equation_experiment(sp.simplify(expr),iv,dv)
```

But it would also be nice to not have to pass it variables, if that's possible at all.

----------------------------------

Further, this method doesn't work with the state functionality yet it seems, but I assume you know that.

**Comment by chadcwilliams** [comment]
This is what I was thinking:

```
import sympy as sp
def equation_runner(expr: str, conditions: pd.DataFrame, added_noise = 0.1):
    equation = sp.simplify(expr)
    x = conditions["x"]
    
    y = [(equation.evalf(subs={equation.args[0]:xi}) + np.random.normal(0, added_noise, size=1))[0] for xi in x]
    return conditions.assign(y=y)

equation_runner('sin(x)', pd.DataFrame({'x':[0,1,3]}))
```

**Comment by chadcwilliams** [comment]
We decided to use https://github.com/AutoResearch/autora-synthetic-abstract-equation.
